// eslint-plugin-monorepo/lib/rules/no-internal-import.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.create = exports.meta = undefined;

var _slicedToArray = function () { function sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"]) _i["return"](); } finally { if (_d) throw _e; } } return _arr; } return function (arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }; }();

var _moduleVisitor = require('eslint-module-utils/moduleVisitor');

var _moduleVisitor2 = _interopRequireDefault(_moduleVisitor);

var _parsePackageName = require('parse-package-name');

var _parsePackageName2 = _interopRequireDefault(_parsePackageName);

var _getMonorepoPackages = require('get-monorepo-packages');

var _getMonorepoPackages2 = _interopRequireDefault(_getMonorepoPackages);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _minimatch = require('minimatch');

var _minimatch2 = _interopRequireDefault(_minimatch);

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var meta = exports.meta = {
  schema: [(0, _moduleVisitor.makeOptionsSchema)({})]
};

var withoutExtension = function withoutExtension(importFile, fileEntry) {
  var importExt = _path2.default.extname(importFile);
  if (importExt !== '') return [importFile, fileEntry];

  var fileEntryExt = _path2.default.extname(fileEntry);
  var newFileEntry = fileEntryExt !== '' ? fileEntry.replace(new RegExp('\\' + fileEntryExt + '$'), '') : fileEntry;
  return [importFile, newFileEntry];
};

var create = exports.create = function create(context) {
  var _context$options = _slicedToArray(context.options, 1),
      moduleUtilOptions = _context$options[0];

  var packages = (0, _getMonorepoPackages2.default)(process.cwd());

  return (0, _moduleVisitor2.default)(function (node) {
    var _tryParse = tryParse(node.value),
        name = _tryParse.name,
        internalPath = _tryParse.path;

    var matchedPackage = packages.find(function (pkg) {
      return pkg.package.name === name;
    });

    if (!internalPath) return;
    if (!matchedPackage) return;

    var packageRoot = matchedPackage.location;

    // Need to take care of "files" field, since they are
    // supposed to be part of the public API of the package
    var absolutePathsForFiles = matchedPackage.package.files && matchedPackage.package.files.reduce(function (acc, file) {
      var fileOrDirOrGlob = _path2.default.join(packageRoot, file);
      acc.push(fileOrDirOrGlob);

      try {
        if (_fs2.default.lstatSync(fileOrDirOrGlob).isDirectory()) {
          // Need to also include
          // all of the files inside the folder
          acc.push(_path2.default.join(fileOrDirOrGlob, '**', '*'));
        }
      } catch (e) {
        // do nothing
      }

      return acc;
    }, []);
    var absoluteInternalPath = _path2.default.join(packageRoot, internalPath);

    if (absolutePathsForFiles) {
      var isImportWithinFiles = absolutePathsForFiles.some(function (maybeGlob) {
        // If import doesn't have an extension, strip it from the file entry
        var _withoutExtension = withoutExtension(absoluteInternalPath, maybeGlob),
            _withoutExtension2 = _slicedToArray(_withoutExtension, 2),
            theImport = _withoutExtension2[0],
            theFileEntry = _withoutExtension2[1];

        return (0, _minimatch2.default)(theImport, theFileEntry);
      });

      if (isImportWithinFiles) return;
    }

    context.report({
      node: node,
      message: 'Import for monorepo package \'' + name + '\' is internal.'
    });
  }, moduleUtilOptions);
};

var tryParse = function tryParse(text) {
  try {
    return (0, _parsePackageName2.default)(text);
  } catch (error) {
    return { path: text, name: '' };
  }
};
